Fix text selection handle alignment, cursor width updates, and Android handle auto-dismissal like native approach

muhammadkamel opened the pull request:
This is a new PR to replace the previously closed PR: https://github.com/flutter/flutter/pull/169715

The text selection anchor/handle in InputField was slightly misaligned, shifting slightly to the left. This PR fixes the handle positioning to ensure proper alignment with the text cursor.

Cursor Width Updates
Previously, updating cursorWidth in EditableText did not reflect immediately without a hot restart. This PR updates EditableTextState.didUpdateWidget to detect cursorWidth changes and rebuild the selection overlay.

Android Handle Auto-Dismissal
On Android, text selection handles now automatically dismiss after 4 seconds of inactivity (specifically when the selection is collapsed), matching native Android behavior. A strict unit test ensures this timeout is respected.


Before:
![baloon_before](https://github.com/user-attachments/assets/99d69ea0-9559-4b79-8402-aa7657dc47b9)


After fix:
![baloon_after_fixed](https://github.com/user-attachments/assets/d89d6a99-f026-4e59-bb17-ad8950aea6f0)

Fixes #45556 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

muhammadkamel committed:
Adjust handle anchor offset for collapsed text selection in MaterialTextSelectionControls

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
update: ancher offset

muhammadkamel committed:
Update text_selection.dart

Updated the calculation of the anchor point for TextSelectionHandleType.collapsed to properly account for the cursor width.

Changed from:
Offset(_kHandleSize / 2, -4.6)
to:
Offset((_kHandleSize - cursorWidth) / 2, -4.6)

This ensures that the anchor aligns with the actual visual center of the cursor.
This change is necessary because the total handle size is 48x48 pixels, which is larger than the actual cursor width (2px), so centering must be adjusted accordingly.

muhammadkamel committed:
Add test for getHandleAnchor in MaterialTextSelectionControls

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge remote-tracking branch 'origin/master' into fix/text-selection-android-anchor

muhammadkamel committed:
feat: add cursorWidth parameter to getHandleAnchor method in text selection controls

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'flutter:master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'flutter:master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge remote-tracking branch 'origin' into fix/text-selection-android-anchor

muhammadkamel committed:
Refactor text selection handling and improve test reliability

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'flutter:master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'flutter:master' into fix/text-selection-android-anchor

muhammadkamel committed:
feat: add required `cursorWidth` parameter to `getHandleAnchor` method.

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
fix: Adjust text selection handle y-offsets and update related test tap positions.

muhammadkamel committed:
fix: ensure SelectionOverlay updates its cursorWidth when EditableText's cursorWidth changes

muhammadkamel committed:
feat: Add auto-dismissal for collapsed text selection handles on Android and corresponding tests.

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
chore[test]: add more tests for centering handle on Android.

muhammadkamel committed:
Merge branch 'flutter:master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'flutter:master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
Merge branch 'flutter:master' into fix/text-selection-android-anchor

justinmc commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -447,14 +464,43 @@ class TextSelectionOverlay {
+  }
+
+  // The duration for auto-dismissal of handles on Android.
+  static const Duration _androidHandlesDismissalDuration = Duration(seconds: 4);
```
Could you add a sentence explaining where you got the 4 seconds from? Like "Measured on a physical Pixel 9 running Android 16." or whatever.

justinmc commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -1254,6 +1308,11 @@ class SelectionOverlay {
 
+  /// Cursor width
+  /// This value is used for calculating the position of the text selection
+  /// handles.
```
Period after "width", and also add an empty line of "///".

justinmc commented on packages/flutter/test/widgets/text_selection_test.dart:
```diff
@@ -1899,6 +1940,54 @@ void main() {
+    expect(handleCenter.dx, closeTo(caretCenterGlobal.dx, 0.5));
+
+    // Also verify specifically that the visual center of the handle (which is the "onion" shape) points to the caret.
+    // The handle logic in getHandleAnchor should have compensated for the handle size and cursor width.
```
Did you intend to add more code to verify this?

justinmc reviewed (commented):
Thank you for being persistent with this change! I think it looks like a great improvement overall. We'll have to make sure that it doesn't break any customers. Can you fix the analyzer error?

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
Improve comments for text selection handling and test clarity

muhammadkamel committed:
Clarify cursor width documentation in SelectionOverlay

muhammadkamel committed:
Set default cursor width to 2.0 in text selection controls

muhammadkamel committed:
Remove unnecessary ignore directive for implementation imports in main.dart

muhammadkamel committed:
Merge branch 'flutter:master' into fix/text-selection-android-anchor

Renzo-Olivares commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -340,6 +352,7 @@ class TextSelectionOverlay {
     required LayerLink endHandleLayerLink,
     required this.renderObject,
     this.selectionControls,
+    required this.cursorWidth,
```
I think it makes sense to make this required but it might be a breaking change.

Renzo-Olivares commented on packages/flutter/lib/src/material/selection_area.dart:
```diff
@@ -54,6 +54,7 @@ class SelectionArea extends StatefulWidget {
     this.contextMenuBuilder = _defaultContextMenuBuilder,
     this.magnifierConfiguration,
     this.onSelectionChanged,
+    this.cursorWidth = 2.0,
```
I'm a bit weary about adding this to `SelectionArea` since i'm not sure how it will be used. A collapsed selection should show no cursor at all on static text. Though I think the reason it is being added is because `TextSelectionOverlay/SelectionOverlay.cursorWidth` is a required argument.

I see both sides:

It is important for `TextSelectionOverlay/SelectionOverlay.cursorWidth` to be a required argument so it can be taken into account when calculating the handle anchor. Making it nullable would mean if a user does not provide it they may not get an accurate handle anchor. Though in the case of `SelectableRegion` and static text we do not show a cursor with a collapsed selection, so adding this parameter is not very useful at the moment.

I think maybe we should hold off on changes to `SelectionArea` and `SelectableRegion` for now since the cursor is not used with static text. We can do this by keeping `TextSelectionOverlay.cursorWidth` required but making `SelectionOverlay.cursorWidth` nullable.

Renzo-Olivares reviewed (commented):
@muhammadkamel thank you for your patience and contribution! This is a great fix and adds a nice feature and improves native fidelity! I left a few comments regarding the surface of this change and thoughts regarding cursorWidth having a default value. Let me know what you think.

muhammadkamel commented on dev/integration_tests/windowing_test/lib/main.dart:
```diff
@@ -10,6 +10,7 @@ import 'dart:io';
 
 import 'package:flutter/foundation.dart';
 import 'package:flutter/material.dart';
+// ignore: implementation_imports
```
@Renzo-Olivares 

I have done this to fix the flutter analyze warnings
But I can revert it if you want

Renzo-Olivares replied:
That's surprising given no other changes in this file. This PR may just need a rebase to master.

muhammadkamel committed:
Address review feedback for text selection handle alignment

- Remove unrelated windowing_test change (implementation_imports comment)
- Add doc template for cursorWidth with default value documentation
- Use doc template reference in TextSelectionOverlay and SelectionOverlay
- Remove cursorWidth from SelectionArea and SelectableRegion since
  collapsed selections don't show cursors on static text
- Make SelectionOverlay.cursorWidth nullable while keeping
  TextSelectionOverlay.cursorWidth required for editable text
- Update test files to reflect nullable cursorWidth in SelectionOverlay

muhammadkamel commented:
@Renzo-Olivares Thank you for the thorough review!

Could you please review new changes now.

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
- Remove cursorWidth parameter from TextSelectionOverlay constructor
- Add cursorWidth getter that reads from renderObject.cursorWidth
- Update SelectionOverlay to receive cursorWidth from renderObject
- Remove cursorWidth from EditableText._createSelectionOverlay call
- Update tests to remove cursorWidth parameter

Renzo-Olivares commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -1254,6 +1313,12 @@ class SelectionOverlay {
+  ///
+  /// This value is used for calculating the position of the text selection
+  /// handles.
+  final double? cursorWidth;
```
This should have a getter and setter similar to `SelectionOverlay.lineHeightAtStart`, this way `TextSelectionOverlay` can update it in `TextSelectionOverlay._updateSelectionOverlay`.

Renzo-Olivares commented on packages/flutter/test/widgets/text_selection_test.dart:
```diff
@@ -1859,6 +1868,36 @@ void main() {
+          ),
+        );
+
+        // Show handles
```
nit: end comments with a period, here and elsewhere in this file.

Renzo-Olivares reviewed (commented):
@muhammadkamel thank you for the quick changes left a few more comments. In general this change is looking good, but I still have some concerns over the naming of `cursorWidth`. It feels strange to me that `SelectionOverlay` has a `cursorWidth` field when it does not paint or control the cursor. Maybe `handleAnchorOffset` is appropriate here, what do you think? Open to other names.

muhammadkamel committed:
Refactor: Rename `cursorWidth` to `handleAnchorOffset` in text selection overlay and handle components to clarify its usage for handle positioning.

muhammadkamel commented on packages/flutter/test/widgets/text_selection_test.dart:
```diff
@@ -2056,6 +2148,11 @@ class FakeRenderEditable extends RenderEditable {
   bool hasFocus = false;
 
+  @override
+  Rect? getRectForComposingRange(TextRange range) {
```
Ohh, I will remove it!

muhammadkamel committed:
Refactor handleAnchorOffset in SelectionOverlay and cleanup tests

- Replaces the `handleAnchorOffset` getter/setter with `updateHandleAnchorOffset` in `SelectionOverlay` to make state updates explicit.
- Updates `TextSelectionOverlay` to use the new method.
- Removes the "Extra large cursor width draws the collapsed handle centered" test from `text_selection_test.dart` which is likely no longer relevant or flaky.
- specific comment punctuation fixes in `editable_text_test.dart`.

Renzo-Olivares reviewed (commented):
Hi @muhammadkamel, thank you for the quick changes. I've been thinking more about the naming around `cursorWidth` and `handleAnchorOffset`. I think a more appropriate name would be `targetWidth`, and some documentation could be as follows:

```dart
/// The [targetWidth] represents the width of the object (e.g. the cursor)
/// that the handle is pointing to. This can be used to center the handle
/// horizontally on that object.
```

Curious what other think of this name, my main concern with the naming (`cursorWidth`) is that this API is being added to `SelectionOverlay` which is used by `SelectableRegion` and does not currently have a visual representation for a collapsed selection because in the framework it only supports static text.

My other thought is around having it default to 2.0 (the same as the default EditableText.cursorWidth). While this matches the behavior in `EditableText`, if there was some other input field implementation with a different cursor width then the fix would not work and the developer would still need to figure out why their handles are not aligned. If it is null by default then that user will need to discover the fix. If it is required then the user will be made aware automatically but this is a breaking change. I'm leaning towards the breaking change so users are notified automatically about having to provide this new member but curious what others think. cc @loic-sharma @justinmc

muhammadkamel committed:
Refactor: convert updateHandleAnchorOffset to public setter
Changes the updateHandleAnchorOffset() method to a public getter/setter.

justinmc commented:
@muhammadkamel Would you be able to fix the conflicts here? Also I owe you a review, sorry for the delay.

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel committed:
docs: clarify cursorWidth and handleAnchorOffset doc comments in TextSelectionControls

dkwingsmt commented:
Hi! @muhammadkamel  Are you able to address the comments above so we can work together to get this landed? This is a great addition and I'm looking forward to it!

muhammadkamel committed:
Merge branch 'flutter:master' into fix/text-selection-android-anchor

muhammadkamel committed:
Deprecate `getHandleAnchor` in favor of `calculateHandleAnchor` and update all related implementations and tests to use the new method. The `calculateHandleAnchor` method includes a required `targetWidth` parameter for better accuracy in handle positioning.

Renzo-Olivares commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -114,7 +114,33 @@ abstract class TextSelectionControls {
+    double textLineHeight, {
+    required double targetWidth,
+  }) {
+    return getHandleAnchor(type, textLineHeight);
```
I don't think this method needs a default implementation. This can be left to the implementer.

Renzo-Olivares commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -114,7 +114,33 @@ abstract class TextSelectionControls {
+  /// that the handle is pointing to. This can be used to center the handle
+  /// horizontally on that object.
+  ///
+  /// Subclasses should override either this method or [getHandleAnchor].
```
I would move this last bit out of the template.

Renzo-Olivares commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -447,14 +477,45 @@ class TextSelectionOverlay {
+  // (e.g., Pixel 9 running Android 16).
+  static const Duration _androidHandlesDismissalDuration = Duration(seconds: 4);
+
+  void _scheduleHandlesDismissal() {
```
nit: maybe `_scheduleHandleDismissalIfSupported`.

Renzo-Olivares commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -1083,10 +1152,12 @@ class SelectionOverlay {
     )
     Offset? toolbarLocation,
     this.magnifierConfiguration = TextMagnifierConfiguration.disabled,
+    required double targetWidth,
```
This is technically a breaking change. 

Option 1: Make this nullable.

Option 2: Add a default value.

If a user:
1. Sets a different cursor width than the default for the area that uses `SelectionOverlay`.
2. Uses `MaterialTextSelectionControls` or another `TextSelectionControls` that uses `targetWidth`.
Then their handle anchor may not be accurately calculated if the user does not know to provide an accurate targetWidth to `SelectionOverlay`.

In both options the handle anchor is left inaccurate if the user does not know to provide an accurate targetWidth to `SelectionOverlay`. I'm leaning towards making this nullable since this was previously not working before and also certain "selectable" widgets like static text do not have a collapsed cursor so they do not need this, also Cupertino handles do not consider `targetWidth`. I don't feel too strongly about either option, ideally users would be notified someway that they need to provide targetWidth after this change but that is a breaking change. Curious to hear others thoughts on this.

By making it required users are broken and notified that they must provide an accurate targetWidth so their handle anchor is properly calculated.

If a user is creating their own `TextSelectionControls` then from implementing `calculateHandleAnchor` they would know to provide an accurate `targetWidth` to `SelectionOverlay`.

Renzo-Olivares reviewed (commented):
Hi @muhammadkamel, thank you for coming back to this and addressing the feedback! I left a few comments.

muhammadkamel commented:
> Hi @muhammadkamel Are you able to address the comments above so we can work together to get this landed? This is a great addition and I'm looking forward to it!

Hi @dkwingsmt 

Sorry for the delay — I was on vacation. I’ll address the comments and keep you updated.

muhammadkamel committed:
Refactor text selection handle methods to use `calculateHandleAnchor` with a required `targetWidth` parameter, and update related tests for consistency.

muhammadkamel committed:
Deprecate `getHandleAnchor` methods in favor of `calculateHandleAnchor` for improved handle positioning accuracy

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

dkwingsmt commented:
(From triage) FYI: Let us know when this is good for another round of review. Thanks!

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

muhammadkamel commented:
> (From triage) FYI: Let us know when this is good for another round of review. Thanks!

Hi @dkwingsmt 

Could you please review the latest commit changes?

muhammadkamel committed:
Merge branch 'flutter:master' into fix/text-selection-android-anchor

Renzo-Olivares commented on packages/flutter/lib/src/cupertino/text_selection.dart:
```diff
@@ -151,11 +151,12 @@ class CupertinoTextSelectionControls extends TextSelectionControls {
     }
   }
 
-  /// Gets anchor for cupertino-style text selection handles.
```
We should re-add this documentation snippet to the deprecated API and the new API as well.

Renzo-Olivares commented on packages/flutter/lib/src/cupertino/desktop_text_selection.dart:
```diff
@@ -68,12 +68,28 @@ class CupertinoDesktopTextSelectionControls extends TextSelectionControls {
     return const SizedBox.shrink();
   }
 
-  /// Gets the position for the text selection handles, but desktop has none.
```
We should re-add this documentation snippet to the deprecated API and the new API as well.

Renzo-Olivares commented on packages/flutter/lib/src/cupertino/text_selection.dart:
```diff
@@ -178,6 +179,19 @@ class CupertinoTextSelectionControls extends TextSelectionControls {
+  Offset getHandleAnchor(
+    TextSelectionHandleType type,
+    double textLineHeight, {
+    double cursorWidth = 2.0,
```
The deprecated API should not have `cursorWidth`. That is reserved for the new API. Also we changed `cursorWidth` -> `targetWidth`.

Renzo-Olivares commented on packages/flutter/lib/src/cupertino/desktop_text_selection.dart:
```diff
@@ -68,12 +68,28 @@ class CupertinoDesktopTextSelectionControls extends TextSelectionControls {
+  Offset getHandleAnchor(
+    TextSelectionHandleType type,
+    double textLineHeight, {
+    double cursorWidth = 2.0,
```
The deprecated API should not have `cursorWidth`. That is reserved for the new API. Also we changed `cursorWidth` -> `targetWidth`.

Renzo-Olivares commented on packages/flutter/lib/src/material/desktop_text_selection.dart:
```diff
@@ -69,12 +69,28 @@ class DesktopTextSelectionControls extends TextSelectionControls {
+  Offset getHandleAnchor(
+    TextSelectionHandleType type,
+    double textLineHeight, {
+    double cursorWidth = 2.0,
```
The deprecated API should not have `cursorWidth`. That is reserved for the new API. Also we changed `cursorWidth` -> `targetWidth`.

Renzo-Olivares commented on packages/flutter/lib/src/material/desktop_text_selection.dart:
```diff
@@ -69,12 +69,28 @@ class DesktopTextSelectionControls extends TextSelectionControls {
     return const SizedBox.shrink();
   }
 
-  /// Gets the position for the text selection handles, but desktop has none.
```
We should re-add this documentation snippet to the deprecated API and the new API as well.

Renzo-Olivares commented on packages/flutter/lib/src/material/text_selection.dart:
```diff
@@ -105,18 +105,32 @@ class MaterialTextSelectionControls extends TextSelectionControls {
     };
   }
 
-  /// Gets anchor for material-style text selection handles.
```
We should re-add this documentation snippet to the deprecated API and the new API as well.

Renzo-Olivares commented on packages/flutter/lib/src/material/text_selection.dart:
```diff
@@ -105,18 +105,32 @@ class MaterialTextSelectionControls extends TextSelectionControls {
+  Offset getHandleAnchor(
+    TextSelectionHandleType type,
+    double textLineHeight, {
+    double cursorWidth = 2.0,
```
The deprecated API should not have cursorWidth. That is reserved for the new API. Also we changed cursorWidth -> targetWidth.

Renzo-Olivares commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -114,7 +114,33 @@ abstract class TextSelectionControls {
+  ///
+  /// Subclasses should override either this method or [getHandleAnchor].
+  /// {@endtemplate}
+  Offset calculateHandleAnchor(
```
A way around this would be to make this method a stub..

Renzo-Olivares commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -111,10 +115,73 @@ abstract class TextSelectionControls {
-  /// Get the anchor point of the handle relative to itself. The anchor point is
-  /// the point that is aligned with a specific point in the text. A handle
-  /// often visually "points to" that location.
-  Offset getHandleAnchor(TextSelectionHandleType type, double textLineHeight);
```
We should re-add this documentation snippet to the deprecated API.

Renzo-Olivares commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -526,20 +665,23 @@ class TextSelectionOverlay {
   /// next frame (i.e. many milliseconds later).
-  void update(TextEditingValue newValue) {
-    if (_value == newValue) {
+  void update(TextEditingValue newValue, {double? targetWidth}) {
```
I recommend exposing a setter for `targetWidth` on `TextSelectionOverlay` rather than changing `update`.

Renzo-Olivares commented:
Since `Material` and `Cupertino` libraries are currently under a code freeze I will mark this as a draft. This will be migrated over when the code freeze ends https://github.com/flutter/flutter/issues/184093.

Piinks commented:
I've marked this PR as not ready to port to flutter/packages yet.
We'll provide instructions to move this change over to material_ui/cupertino_ui once ready to receive PRs. Thank you!

The changes not in the material or cupertino libraries will remain here.

Piinks commented:
Advice for PR splitting: Since this touches material/cupertino along with other parts of flutter/flutter, it will need to be split up.

If the material/cupertino changes are dependent on the other non-material/cupertino changes in this PR, then the material/cupertino changes will need to wait for them to roll to stable before they can be ported over to material_ui/cupertino_ui in flutter/packages (which uses the stable branch).
- If that is the case here, consider opening a new PR with the non-material/cupertino changes to move forward with so that waiting period can be reduced. Up to you!

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-android-anchor

Piinks commented:
If you decide to return to this, material_ui and cupertino_ui are now accepting contributions. :)

The portions of this PR that affect the material and cupertino libraries can be ported over to flutter/packages, although they may have to wait for the changes here in the core SDK to reach stable first.

Guidance and instructions for easily porting over can be found in https://github.com/flutter/flutter/issues/188444
